### PR TITLE
3.0.0-alpha6: "openssl pkcs12" is unable to parse or create PKCS#12 archives with default ciphers

### LINKED ISSUE #11672: 3.0.0-alpha1: "openssl pkcs12" is unable to parse or create PKCS#12 archives with default ciphers
Works in 1.1.1:
```
$ openssl version
OpenSSL 1.1.1d FIPS  10 Sep 2019
$ openssl pkcs12 -in cert.pem -inkey key.pem -out foo.p12 -export
Enter Export Password:
Verifying - Enter Export Password:
```

Fails in 3.0.0:
```
$ $HOME/openssl-3.0.0/bin/openssl version
OpenSSL 3.0.0-alpha1 23 Apr 2020 (Library: OpenSSL 3.0.0-alpha1 23 Apr 2020)
$ $HOME/openssl-3.0.0/bin/openssl pkcs12 -in cert.pem -inkey key.pem -out foo.p12 -export
Enter Export Password:
Verifying - Enter Export Password:
40:A7:D6:79:1B:7F:00:00:error:digital envelope routines:evp_generic_fetch:fetch failed:crypto/evp/evp_fetch.c:307:Default library context, Algorithm (RC2-40-CBC), Properties ()
40:A7:D6:79:1B:7F:00:00:error:digital envelope routines:EVP_CipherInit_ex:initialization error:crypto/evp/evp_enc.c:336:
40:A7:D6:79:1B:7F:00:00:error:digital envelope routines:EVP_PBE_CipherInit:keygen failure:crypto/evp/evp_pbe.c:132:
40:A7:D6:79:1B:7F:00:00:error:PKCS12 routines:PKCS12_pbe_crypt:pkcs12 algor cipherinit error:crypto/pkcs12/p12_decr.c:36:
40:A7:D6:79:1B:7F:00:00:error:PKCS12 routines:PKCS12_item_i2d_encrypt:encrypt error:crypto/pkcs12/p12_decr.c:133:
40:A7:D6:79:1B:7F:00:00:error:PKCS12 routines:PKCS12_pack_p7encdata:encrypt error:crypto/pkcs12/p12_add.c:119:
```

### PR DESCRIPTION
**Pull request by sahanaprasad07**
This PR makes the following two changes in openssl-pkcs12 to fix #11672 :

1. Changes the default encryption algorithms for certificates and private keys from RC2 to AES_256_CBC in OpenSSL 3.0
2. Add a new option "-legacy" to the pkcs12 cli which loads the legacy provider and restores the default algorithms
    The default algorithm for private key encryption is 3DES_CBC. If the legacy option is not specified, then 
     the legacy provider is not loaded and the default encryption algorithm for both certificates and private keys is
    AES_256_CBC by default.

The above design was proposed by @t8m as per the comments in the discussion #11672 that were agreed upon.

Documentation and example section is updated.
No new tests are added. The cli usage is documented under examples.

Usage:

openssl pkcs12 -in cert.pem -inkey key.pem -out foo.p12 -export
openssl pkcs12 -info -in foo.p12

with the legacy option:

openssl pkcs12 -in cert.pem -inkey key.pem -out foo.p12 -export **-legacy**
openssl pkcs12 -info -in foo.p12 **-legacy**

Note: certs and keys can be generated using the below command
openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 365

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by sahanaprasad07** [comment]
closing this PR and re-opening it again so that CLA flag gets updated.

**Comment by sahanaprasad07** [comment]
Had closed it previously as cla flag was not updated.

File: apps/pkcs12.c [hunk 1]
```diff
@@ -311,13 +308,38 @@ int pkcs12_main(int argc, char **argv)
...
+    if (use_legacy) {
+        /* load the legacy provider if not loaded already*/
+        if (!OSSL_PROVIDER_available(app_get0_libctx(), "legacy")) {
+            if (!app_provider_load(app_get0_libctx(), "legacy"))
+                goto end;
+        }
+
+        /* load the default provider explicitly */
+        if (!app_provider_load(app_get0_libctx(), "default"))
+            goto end;
```
Comment (t8m) [inline apps/pkcs12.c:331]: This should probably be moved inside the previous if block.

File: doc/man1/openssl-pkcs12.pod.in [hunk 2]
```diff
@@ -58,6 +58,7 @@ B<openssl> B<pkcs12>
 {- $OpenSSL::safe::opt_r_synopsis -}
 {- $OpenSSL::safe::opt_engine_synopsis -}
 {- $OpenSSL::safe::opt_provider_synopsis -}
+[B<-legacy>]
```
Comment (t8m) [inline doc/man1/openssl-pkcs12.pod.in:61]: Maybe move this up before the "common" options. And similarly below with the legacy option description.

File: apps/pkcs12.c (hunk 1 above)
Comment (sahanaprasad07) [reply apps/pkcs12.c:333]: default provider won't get loaded then if legacy provided was somehow already found through OSSL_PROVIDER_available() right?

Comment (t8m) [reply apps/pkcs12.c:333]: I would say that unless we load the legacy provider explicitly, we do not want to mess with the provider loading at all. User might have configuration where both legacy and default is loaded already. If he has a (broken) configuration where the default is not loaded alongside the legacy provider then he deserves to get the failure from the missing crypto from the default provider anyway.

File: doc/man1/openssl-pkcs12.pod.in [hunk 3]
```diff
@@ -166,6 +167,16 @@ always assumes these are the same so this option will render such
 PKCS#12 files unreadable. Cannot be used in combination with the options
 B<-password>, B<-passin> if importing, or B<-passout> if exporting.
 
+=item B<-legacy>
+
+This option is used to fallback to the default algorithms as in the legacy
+provider. In the legacy mode, the default algorithm for certificate encryption
```
Comment (t8m) [inline doc/man1/openssl-pkcs12.pod.in:173]: I would change the first sentence to something like: `Use legacy mode of operation and automatically load the legacy provider.`

File: doc/man1/openssl-pkcs12.pod.in [hunk 4]
```diff
@@ -166,6 +167,16 @@ always assumes these are the same so this option will render such
...
+is RC2_CBC or 3DES_CBC with or without the OPENSSL_NO_RC2 option respectively.
```
Comment (t8m) [inline doc/man1/openssl-pkcs12.pod.in:174]: Replace `with or without the OPENSSL_NO_RC2 option respectively.` with `depending on whether the RC2 cipher is enabled in the build.`

File: doc/man1/openssl-pkcs12.pod.in [hunk 5]
```diff
@@ -166,6 +167,16 @@ always assumes these are the same so this option will render such
...
+The default algorithm for private key encryption is 3DES_CBC. If the legacy
+option is not specified, then the legacy provider is not loaded and the
+default encryption algorithm for both certificates and private keys is
+AES_256_CBC by default.
```
Comment (t8m) [inline doc/man1/openssl-pkcs12.pod.in:178]: Please replace `by default` with `with PBKDF2 for key derivation by default.`

File: apps/pkcs12.c [hunk 6]
```diff
@@ -211,7 +208,7 @@ int pkcs12_main(int argc, char **argv)
             macver = 0;
             break;
         case OPT_DESCERT:
-            cert_pbe = NID_pbe_WithSHA1And3_Key_TripleDES_CBC;
+            cert_pbe = NID_aes_256_cbc;
```
Comment (slontis) [inline apps/pkcs12.c:211]: How is this related to des?

File: apps/pkcs12.c [hunk 7]
```diff
@@ -313,13 +310,37 @@ int pkcs12_main(int argc, char **argv)
...
             if (!opt_provider(o))
                 goto end;
             break;
         }
     }
     argc = opt_num_rest();
+
+    if (use_legacy) {
+        /* load the legacy provider if not loaded already*/
+        if (!OSSL_PROVIDER_available(app_get0_libctx(), "legacy")) {
```
Comment (slontis) [inline apps/pkcs12.c:326]: #include <openssl/provider.h> at the top to use this API.

File: apps/pkcs12.c (hunk 6 above)
Comment (slontis) [reply apps/pkcs12.c:211]: IMO it should either fail OR
use des and set use_legacy

Comment (sahanaprasad07) [reply apps/pkcs12.c:211]: @slontis Thanks for your comment. The latter would reset key_pbe as well. Maybe we don't want that. So I could just set it back to - cert_pbe = NID_pbe_WithSHA1And3_Key_TripleDES_CBC here and nothing else. What do you think?

Comment (t8m) [reply apps/pkcs12.c:211]: IMO 3DES (at least for now) should work without legacy provider, shouldn't it? So yes, just dropping this change should suffice.

Comment (t8m) [reply apps/pkcs12.c:211]: It would be worth explicitly documenting that -descert will use 3DES for the cert encryption and AES will be used for key unless -legacy is used. Also use_legacy handling needs to be adjusted to not override the cert_pbe if it is set to  NID_pbe_WithSHA1And3_Key_TripleDES_CBC already.

Comment (sahanaprasad07) [reply apps/pkcs12.c]: > IMO 3DES (at least for now) should work without legacy provider, shouldn't it? So yes, just dropping this change should suffice.

yes it works without legacy provider.

**Comment by sahanaprasad07** [comment]
> A few more doc nits.

Fixed.

**Comment by sahanaprasad07** [comment]
Closed by mistake

**Comment by t8m** [comment]
I've squashed the commits and merged them to master.
Thank you for the PR.

**Comment by levitte** [comment]
Hmmm, should this become a standard option alongside `-provider` and `-provider_path`?  I can't really see how it's justifiable to special case the "pkcs12" command...

**Comment by t8m** [comment]
Do we have clear use cases for this option with other commands? Or do you mean the generalized case to be "just load the legacy and default provider"?

**Comment by t8m** [comment]
The special case here is that we clearly need to change the defaults to something more modern with pkcs12 but the old defaults might be needed for interoperability AND they require the legacy provider. With other commands I do not see urgent need to change the defaults but if you think about any, where it makes sense, I'd agree with the generalization.
